### PR TITLE
feat: store chunk to network as kad record directly

### PR DESCRIPTION
**Pull request by maqi**
This PR contains the work of store chunk to network as kad record directly.
This allows the data replication to be covered by ligp2p directly.

Closes https://github.com/maidsafe/safe_network/pull/172
Closes https://github.com/maidsafe/safe_network/pull/155

Related to Issue https://github.com/maidsafe/safe_network/issues/46

### CODE REVIEW

File: safenode/src/network/mod.rs [hunk 1]
```diff
@@ -296,6 +335,24 @@ impl Network {
...
+        let (sender, receiver) = oneshot::channel();
+        self.send_swarm_cmd(SwarmCmd::GetProvidedData { key, sender })
+            .await?;
+
+        receiver
+            .await
+            .map_err(|_e| Error::InternalMsgChannelDropped)
+    }
+
+    /// Register self as Provider for Data
```
Comment (RolandSherwin) [inline safenode/src/network/mod.rs]: Just a small one, the doc needs to be updated here.

File: safenode/src/node/api.rs [hunk 2]
```diff
@@ -200,12 +202,19 @@ impl Node {
         match query {
             Query::Register(query) => self.registers.read(&query, User::Anyone).await,
             Query::GetChunk(address) => {
-                let resp = self
-                    .chunks
-                    .get(&address)
+                match self
+                    .network
+                    .get_provided_data(RecordKey::new(address.name()))
```
Comment (RolandSherwin) [inline safenode/src/node/api.rs:207]: So, if I'm not wrong, Client sends a `GetChunk` and nodes fetch the data for it? 
If so, maybe we can directly call `network.swarm.behaviour_mut().kademlia.get_record(key);` from the client. This has the advantage of short-circuiting when the data is found. Might speed up the queries by a bit.

Maybe we can do this in the next PR?

File: safenode/src/network/cmd.rs [hunk 3]
```diff
@@ -74,43 +71,16 @@ impl SwarmDriver {
...
-                            .send(crate::protocol::messages::QueryResponse::GetChunk(Err(
-                                ProtocolError::ProvideRecordNotFound,
-                            )))
-                            .map_err(|_| Error::InternalMsgChannelDropped)?;
-
-                        warn!("No data found in kademlia store for key: {:?}", key);
-                    }
-                }
+                let query_id = self.swarm.behaviour_mut().kademlia.get_record(key);
+                let _ = self.pending_query.insert(query_id, sender);
```
Comment (joshuef) [inline safenode/src/network/cmd.rs:76]: So does this not actually the fire the request away into the network again here? 

I'm not sure we're actually getting data from _this node_?

File: safenode/src/network/mod.rs [hunk 4]
```diff
@@ -85,8 +90,27 @@ impl SwarmDriver {
...
+        let _ = cfg.set_replication_factor(
+            NonZeroUsize::new(CLOSE_GROUP_SIZE).ok_or_else(|| Error::InvalidCloseGroupSize)?,
+        );
         let _ = cfg.set_query_timeout(Duration::from_secs(5 * 60));
 
+        // Require iterative queries to use disjoint paths for increased resiliency in the presence of potentially adversarial nodes.
+        let _ = cfg.disjoint_query_paths(true);
+
+        // Provider records never expire
+        let _ = cfg.set_provider_record_ttl(None);
```
Comment (RolandSherwin) [inline safenode/src/network/mod.rs]: I think we might have to set, `cfg.set_record_ttl(None)` to make the normal data `Records` not expire.

File: safenode/src/network/cmd.rs (hunk 3 above)
Comment (maqi) [reply safenode/src/network/cmd.rs:76]: `get_record` will fire the query to network, and a response will be given back later.

File: safenode/src/node/api.rs (hunk 2 above)
Comment (maqi) [reply safenode/src/node/api.rs:207]: > we can directly call network.swarm.behaviour_mut().kademlia.get_record(key); from the client

yes, I think we shall let client do this directly

Comment (maqi) [reply safenode/src/node/api.rs:207]: pushed change that makes client do the `get_record` directly.

Comment (RolandSherwin) [reply safenode/src/node/api.rs:207]: Hey, sorry, do we still need `Query::GetChunk` and `Cmd::StoreChunk` now that the clients talk with the swarm directly and not the peers?

Comment (maqi) [reply safenode/src/node/api.rs:207]: we may still need them for the later on usage, say some data exchanged among nodes.
so, I think we can keep them

**Review by RolandSherwin — APPROVED** [review]

**Pull request merged**